presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KdbTree.java

```diff
@@ -0,0 +1,358 @@
+                .compare(first.getYMin(), second.getYMin())
+                .result();
+
+        @Override
+        public Comparator<Rectangle> getComparator()
+        {
+            return comparator;
+        }
+
+        @Override
+        public double getValue(Rectangle rectangle)
+        {
+            return rectangle.getXMin();
+        }
+
+        @Override
+        public SplitResult<Rectangle> split(Rectangle rectangle, double x)
+        {
+            checkArgument(rectangle.getXMin() < x && x < rectangle.getXMax());
+            return new SplitResult<>(
+                    new Rectangle(rectangle.getXMin(), rectangle.getYMin(), x, rectangle.getYMax()),
+                    new Rectangle(x, rectangle.getYMin(), rectangle.getXMax(), rectangle.getYMax()));
+        }
+    };
+
+    private static final SplitDimension BY_Y = new SplitDimension() {
+        private final Comparator<Rectangle> comparator = (first, second) -> ComparisonChain.start()
+                .compare(first.getYMin(), second.getYMin())
+                .compare(first.getXMin(), second.getXMin())
+                .result();
+
+        @Override
+        public Comparator<Rectangle> getComparator()
+        {
+            return comparator;
+        }
+
+        @Override
+        public double getValue(Rectangle rectangle)
+        {
+            return rectangle.getYMin();
+        }
+
+        @Override
+        public SplitResult<Rectangle> split(Rectangle rectangle, double y)
+        {
+            checkArgument(rectangle.getYMin() < y && y < rectangle.getYMax());
+            return new SplitResult<>(
+                    new Rectangle(rectangle.getXMin(), rectangle.getYMin(), rectangle.getXMax(), y),
+                    new Rectangle(rectangle.getXMin(), y, rectangle.getXMax(), rectangle.getYMax()));
+        }
+    };
+
+    private static final class LeafIdAllocator
+    {
+        private int nextId;
+
+        public int next()
+        {
+            return nextId++;
+        }
+    }
+
+    public static KdbTree buildKdbTree(int maxItemsPerNode, Rectangle extent, List<Rectangle> items)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+        return new KdbTree(buildKdbTreeNode(maxItemsPerNode, 0, extent, items, new LeafIdAllocator()));
+    }
+
+    private static Node buildKdbTreeNode(int maxItemsPerNode, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(level >= 0, "level must be >= 0");
+        checkArgument(level <= MAX_LEVELS, "level must be <= 10,000");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+
+        if (items.size() <= 1 || items.size() <= maxItemsPerNode || level == MAX_LEVELS) {
```

Comment:
if `maxItemsPerNode > 0` (per the check above), then `items.size() <= maxItemsPerNode` => `items.size() <= 1`, so the latter is redundant.